nestkernel/conn_builder.cpp

```diff
@@ -74,75 +71,145 @@ nest::ConnBuilder::ConnBuilder( NodeCollectionPTR sources,
 #endif
 
-  if ( not default_weight_and_delay_ )
-  {
-    weight_ = syn_spec->known( names::weight )
-      ? ConnParameter::create( ( *syn_spec )[ names::weight ], kernel().vp_manager.get_num_threads() )
-      : ConnParameter::create( ( *syn_defaults )[ names::weight ], kernel().vp_manager.get_num_threads() );
-    register_parameters_requiring_skipping_( *weight_ );
-    delay_ = syn_spec->known( names::delay )
-      ? ConnParameter::create( ( *syn_spec )[ names::delay ], kernel().vp_manager.get_num_threads() )
-      : ConnParameter::create( ( *syn_defaults )[ names::delay ], kernel().vp_manager.get_num_threads() );
-  }
-  else if ( default_weight_ )
-  {
-    delay_ = syn_spec->known( names::delay )
-      ? ConnParameter::create( ( *syn_spec )[ names::delay ], kernel().vp_manager.get_num_threads() )
-      : ConnParameter::create( ( *syn_defaults )[ names::delay ], kernel().vp_manager.get_num_threads() );
+    if ( not default_weight_and_delay_[ indx ] )
+    {
+      weight_[ indx ] = syn_params->known( names::weight )
+        ? ConnParameter::create( ( *syn_params )[ names::weight ], kernel().vp_manager.get_num_threads() )
+        : ConnParameter::create( ( *syn_defaults )[ names::weight ], kernel().vp_manager.get_num_threads() );
+      register_parameters_requiring_skipping_( *weight_[ indx ] );
+      delay_[ indx ] = syn_params->known( names::delay )
+        ? ConnParameter::create( ( *syn_params )[ names::delay ], kernel().vp_manager.get_num_threads() )
+        : ConnParameter::create( ( *syn_defaults )[ names::delay ], kernel().vp_manager.get_num_threads() );
+    }
+    else if ( default_weight_[ indx ] )
+    {
+      delay_[ indx ] = syn_params->known( names::delay )
+        ? ConnParameter::create( ( *syn_params )[ names::delay ], kernel().vp_manager.get_num_threads() )
+        : ConnParameter::create( ( *syn_defaults )[ names::delay ], kernel().vp_manager.get_num_threads() );
+    }
+    register_parameters_requiring_skipping_( *delay_[ indx ] );
+
+    for ( Dictionary::const_iterator default_it = syn_defaults->begin(); default_it != syn_defaults->end();
+          ++default_it )
+    {
+      const Name param_name = default_it->first;
+      if ( skip_set.find( param_name ) != skip_set.end() )
+      {
+        continue; // weight, delay or not-settable parameter
+      }
+
+      if ( syn_params->known( param_name ) )
+      {
+        synapse_params_[ indx ][ param_name ] =
+          ConnParameter::create( ( *syn_params )[ param_name ], kernel().vp_manager.get_num_threads() );
+        register_parameters_requiring_skipping_( *synapse_params_[ indx ][ param_name ] );
+      }
+    }
+
+    // Now create dictionary with dummy values that we will use
+    // to pass settings to the synapses created. We create it here
+    // once to avoid re-creating the object over and over again.
+    if ( synapse_params_[ indx ].size() > 0 )
+    {
+      for ( thread tid = 0; tid < kernel().vp_manager.get_num_threads(); ++tid )
```

Comment:
Can we do this thread-parallel?

Reply:
@heplesser I have tried to make it thread-parallel, but for some reason I was not able to. Some of the tests exited with a segmentation fault.

Reply:
Ok, it is not so time critical, let's leave it as it is and return to it if later profiling shows it to be a bottleneck.